src/toil/batchSystems/kubernetes.py

```diff
@@ -1603,40 +1788,47 @@ def killBatchJobs(self, jobIDs: List[int]) -> None:
         # Clears workflow's jobs listed in jobIDs.
 
         # First get the jobs we even issued non-locally
-        issuedOnKubernetes = set(self._getIssuedNonLocalBatchJobIDs())
+        issued_on_kubernetes = set(self._getIssuedNonLocalBatchJobIDs())
+        deleted_jobs: List[str] = []
 
-        for jobID in jobIDs:
+        for job_id in jobIDs:
             # For each job we are supposed to kill
-            if jobID not in issuedOnKubernetes:
+            if job_id not in issued_on_kubernetes:
                 # It never went to Kubernetes (or wasn't there when we just
                 # looked), so we can't kill it on Kubernetes.
+
+                # It is likely that the job is still waiting inside the jobs
+                # queue and hasn't been submitted to Kubernetes. In this case,
+                # it is not reliable to loop through the queue and try to
+                # delete it. Instead, let's keep track of it and don't submit
+                # when we encounter it.
+                self._killed_queue_jobs.add(job_id)
```

Comment:
This (and the `_getIssuedNonLocalBatchJobIDs()` call) needs to be protected with a lock that stops the scheduler from submitting the job to Kubernetes at the same time we are checking to see if it is there and if not adding it to the list of jobs not to submit.

I think you can probably just do it inside a `with self._work_available:`.

Reply:
Also `_killed_queue_jobs` will get arbitrarily big if we kill a lot of jobs that aren't actually running or in the queue (maybe because they finished?). Not sure if there's anything reasonable to do about that though.

Reply:
That is a good catch! I was worried that there may be race conditions that I didn't account for. I think it would make sense to use the same lock as the scheduler for this.

> Also _killed_queue_jobs will get arbitrarily big if we kill a lot of jobs that aren't actually running or in the queue (maybe because they finished?). Not sure if there's anything reasonable to do about that though.

This is also true, but yeah I don't think we can do anything about it? We could check if the job is actually in the queue, but it might just be possible that a job hasn't been added to our queue yet and we end up not considering it.